optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -67,16 +68,44 @@ public boolean isConstraintMatchEnabled() {
     // Worker methods
     // ************************************************************************
 
+    protected class MultiLevelConstraintUndoListener implements ActivationUnMatchListener {
+
+        private Map<Integer, ActivationUnMatchListener> activationUnMatchListenerMap;
+
+        public MultiLevelConstraintUndoListener(int scoreLevel, ActivationUnMatchListener activationUnMatchListener) {
+            this.activationUnMatchListenerMap = new HashMap<Integer, ActivationUnMatchListener>();
```

Comment:
I wondering about the memory impact of creating all these HashMaps.
We have a listener instance for each rule for each match. The HashMap.DEFAULT_INITIAL_CAPACITY is 16.